tests/compiler/std/map.optimized.wat

```diff
@@ -2483,10 +2483,6 @@
     local.get $1
     local.get $0
     local.get $0
-    i32.const 24
-    i32.shl
-    i32.const 24
-    i32.shr_s
```

Comment:
Hmm, do you provide some extra optimizations?

Reply:
Perhaps that's something Binaryen does now. Iirc correctly I didn't update optimized fixtures in https://github.com/AssemblyScript/assemblyscript/pull/911

Reply:
Hmm, can't spot a PR there that might be responsible for this. Needs investigation.

Reply:
That's strange because I can't reproduce this code shrinkage in master rebuilding std/map. Also it looks incorrect and just ignoring wrapping routines for i8/i16. It seems correntFlow loose / rewrite some info about current type

Reply:
The code in question is

```ts
  for (let k: K = 0; k < 100; ++k) {
    assert(!map.has(k));
    map.set(k, 10 + <V>k); // here
```

where it previously sign-extended `k`. What I think happened there is that the compiler previously compiled the incrementor `++k` too early, in turn leading to a "not wrapped" flag on `k` since it can overflow after that operation. With the incrementor now correctly being compiled after the body, the compiler knows that a constant value of `0` is properly wrapped, hence doesn't do sign extension anymore.

Reply:
A correct solution would be to unroll the first iteration of the loop, with the first being based on the side-effects of the initializer and any subsequent iteration being based on the effects of the incrementor. This comes at a code size hit ofc, especially if the body is large, plus there is no infrastructure yet to suppress errors when compiling the body a second time :(

Reply:
Would look like

```wat
(block $break
 (initializer)
 (br_if !condition $break) ;; -> unreachable code detected if always false
 (block $continue-first
  (first-iteration ...)
 )
 (loop $continue
  (incrementor)
  (br_if !condition $break)
  (subsequent-iteration ...)
  (br $continue)
 )
)
```

Reply:
Hmm, split every loop to this definitely bad idea in reason code size & performance. I still can't understand what is problem you try to solve? Why we don't need "workaround" before?

Reply:
While the workaround we had before did work in this specific situation, it is not a proper fix. For instance, if the initializer sets a non-wrapped state and the incrementor sets a wrapped state, then it's the other way around and will still be wrong.

Reply:
An alternative might be to implement a mechanism in flows that can determine the least common denominator of two flows so when we compile the body, this common state is used. This reduces code size, but may do unnecessary wrapping in either the first or the subsequent iterations.

Reply:
that's better. Unnecessary wrapping could eliminate by binaryen's peephole optimizer